rosidl_typesupport_opendds_cpp/resource/srv__type_support.cpp.em

```diff
@@ -262,8 +342,38 @@ bool take_request__@(service.namespaced_type.name)(
   rmw_request_id_t * request_header,
   void * untyped_ros_request)
 {
-@# TODO: Implement, considering original code in ffe10f9 or earlier
-  return true;
+  using ROSRequestType = @(__ros_request_msg_type);
+  using ReplierType = rosidl_typesupport_opendds_cpp::Replier<
+    @(__dds_msg_typesupport_type)RequestWrapper,
+    @(__dds_msg_typesupport_type)ResponseWrapper
+  >;
+  if (!untyped_replier || !request_header || !untyped_ros_request) {
+    return false;
+  }
+  
+  ReplierType * replier = static_cast<ReplierType *>(untyped_replier);
+
+  @(__dds_msg_typesupport_type)RequestWrapper dds_request_wrapper;
+
+  if (DDS::RETCODE_OK != replier->take_request(dds_request_wrapper)) {
+    RMW_SET_ERROR_MSG("take_request failed");
+    return false;
+  }
+
+  int64_t sequence_number =
+    (((int64_t)dds_request_wrapper.header().request_id().sequence_number().high) << 32) |
+    dds_request_wrapper.header().request_id().sequence_number().low;
+  request_header->sequence_number = sequence_number;
+
+  size_t IDENTITY_SIZE = 16;
+  OpenDDS::DCPS::GUID_t id = dds_request_wrapper.header().request_id().writer_guid();
+  std::memcpy(&(request_header->writer_guid[0]), &id, IDENTITY_SIZE);
+
+  ROSRequestType* ros_request = static_cast<ROSRequestType *>(untyped_ros_request);
```

Comment:
smart pointer vs raw pointer?

Reply:
No ownership is happening here, it's just looking at the data to convert it.